src/server/mod.rs

```diff
@@ -273,73 +279,97 @@ impl<H: Handler + 'static> Worker<H> {
 
         // FIXME: Use Type ascription
         let stream_clone: &mut NetworkStream = &mut stream.clone();
-        let rdr = BufReader::new(stream_clone);
-        let wrt = BufWriter::new(stream);
+        let mut rdr = BufReader::new(stream_clone);
+        let mut wrt = BufWriter::new(stream);
 
-        self.keep_alive_loop(rdr, wrt, addr);
+        while self.keep_alive_loop(&mut rdr, &mut wrt, addr) {
+            if let Err(e) = self.set_read_timeout(rdr.get_mut(), self.timeouts.keep_alive) {
+                error!("set_read_timeout keep_alive {:?}", e);
+                break;
+            }
+        }
 
         self.handler.on_connection_end();
 
         debug!("keep_alive loop ending for {}", addr);
     }
 
+    fn set_timeouts(&self, s: & &mut NetworkStream) -> io::Result<()> {
```

Comment:
The double indirection is crappy, but using a generic wasn't really possible, since the type is erased into a trait object inside the keep alive loop. And since `NetworkStream` requires `'static`, I couldn't implement the trait for `&mut T`. Fun.